### PR TITLE
Migrate tests and fix query cache stats, native query variable substitution and named query support as loader

### PR DESCRIPTION
**Pull request by beikov**

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/dialect/OracleDialect.java [hunk 1]
```diff
@@ -643,6 +646,17 @@ public void contributeTypes(TypeContributions typeContributions, ServiceRegistry
 
 			typeContributions.contributeSqlTypeDescriptor( descriptor );
 		}
+
+		// We need a special type descriptor for Oracle, because that must return BIT as SQL type
+		// in order for PreparedStatement#setNull to work properly
+		typeContributions.contributeSqlTypeDescriptor(
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/dialect/OracleDialect.java:652]: I really think we should try `Types#NULL` for the cases where we need to set null values.  No idea if that works for all drivers/databases but if so its a much better option IMO than randomly picking one

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/dialect/OracleDialect.java:652]: The other option is to delegate binding nulls to the `WrapperOptions` object. It's what it exists for.

We could either let `WrapperOptions` remap "null" JDBC type codes, or we could let it actually handle the whole call to `setNull()`.

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/dialect/OracleDialect.java:652]: Or even delegate back to Dialect through WrapperOptions

Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/dialect/OracleDialect.java:652]: Yep, right, that's what I mean. Should have been more explicit.

File: hibernate-core/src/main/java/org/hibernate/dialect/OracleDialect.java [hunk 2]
```diff
@@ -643,6 +653,79 @@ public void contributeTypes(TypeContributions typeContributions, ServiceRegistry
 
 			typeContributions.contributeSqlTypeDescriptor( descriptor );
 		}
+		typeContributions.contributeSqlTypeDescriptor( new OracleBooleanTypeDescriptor() );
+	}
+
+	private static final class OracleBooleanTypeDescriptor extends BooleanTypeDescriptor {
+
+		private static final Logger log = CoreLogging.logger( BasicBinder.class );
+
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/dialect/OracleDialect.java:662]: We should probably instead create a "subsystem" logger here and use that from the various impls.  More just a mental todo...

**Review by sebersole — APPROVED** [review]
Awesome!

**Comment by sebersole** [comment]
I say squash and push

**Comment by sebersole** [comment]
Nice!  And just in time :)

**Pull request merged**